packages/vc-export/src/KiltAttestationProofV1.ts

```diff
@@ -0,0 +1,668 @@
+    statement,
+  }))
+  const sorted = statementsAndDigests.sort((a, b) => {
+    return u8aCmp(a.digest, b.digest)
+  })
+  return {
+    statements: sorted.map(({ statement }) => statement),
+    digests: sorted.map(({ digest }) => digest),
+  }
+}
+
+function makeCommitments(
+  statementDigests: Uint8Array[],
+  salt: Uint8Array[]
+): Uint8Array[] {
+  return statementDigests.map((digest, index) => {
+    // initialize array with 36 + 2 + 64 bytes
+    const bytes = new Uint8Array(102)
+    // add salt to array
+    bytes.set(salt[index])
+    // add bytes 0x30 & 0x78
+    bytes.set([48, 120], 36)
+    // add hex encoded digest
+    bytes.set(stringToU8a(u8aToHex(digest, undefined, false)), 38)
+    // compute commitment
+    return blake2AsU8a(bytes, 256)
+  })
+}
+
+/**
+ * (Re-)computes the root hash / credential hash from a credential and proof.
+ *
+ * @param credential A [[KiltCredentialV1]] type credential.
+ * @param proof A [[KiltAttestationProofV1]] type proof for this credential.
+ * @returns The root hash.
+ */
+export function calculateRootHash(
+  credential: Pick<VerifiableCredential, 'federatedTrustModel'> &
+    Partial<VerifiableCredential>,
+  proof: Pick<KiltAttestationProofV1, 'commitments'> &
+    Partial<KiltAttestationProofV1>
+): Uint8Array {
+  const { federatedTrustModel = [] } = credential
+  const { commitments } = proof
+  const rootHashInputs = [
+    // Collect commitments for root hash
+    ...commitments.map((i) => base58Decode(i)),
+    // Collect trust model items for root hash
+    ...federatedTrustModel.map((entry) => {
+      if (entry.type === KILT_ATTESTER_LEGITIMATION_V1_TYPE) {
+        // get root hash from credential id
+        return credentialIdToRootHash(entry.id)
+      }
+      if (entry.type === KILT_ATTESTER_DELEGATION_V1_TYPE) {
+        // get on-chain id from delegation id
+        return delegationIdFromAttesterDelegation(entry)
+      }
+      throw new CredentialMalformedError(
+        `unknown type ${
+          (entry as { type: string }).type
+        } in federatedTrustModel`
+      )
+    }),
+  ]
+  // Concatenate and hash to produce root hash
+  return blake2AsU8a(u8aConcatStrict(rootHashInputs), 256)
+}
+
+async function verifyAttestedAt(
+  claimHash: Uint8Array,
+  blockHash: Uint8Array,
+  opts: { api?: ApiPromise } = {}
+): Promise<{
+  verified: boolean
+  timestamp: number
+  attester: DidUri
+  cTypeId: ICType['$id']
+  delegationId: IDelegationNode['id'] | null
+}> {
+  const { api = ConfigService.get('api') } = opts
+  const apiAt = await api.at(blockHash)
+  // TODO: should we look up attestation storage as well and take attestation info from there? That gives us the definitive attestation state in this block (e.g. makes sure it hasn't been removed after)
+  const [events, time] = await apiAt.queryMulti<
+    [Vec<FrameSystemEventRecord>, u64]
+  >([
+    [api.query.system.events as QueryableStorageEntry<'promise'>],
```

Comment:
Wouldn't it be easier to simply query the attestation storage instead? What was the reason for parsing the block events instead?

Reply:
The attestation storage does not tell us when the attestation happened. Therefore we check the event, and check the storage only when checking revocation status.